File: harvester/parsers/registry/engine.py
"""Parse engine that orchestrates element parsing using the registry."""

from __future__ import annotations

from typing import TYPE_CHECKING

from harvester.parsers.registry.protocols import ParseContext, ParseResult
from harvester.parsers.registry.registry import ElementRegistry, get_tag_name

if TYPE_CHECKING:
    from lxml import etree


class UnknownElementError(Exception):
    """Raised when encountering an element without a registered handler."""

    def __init__(self, tag_name: str, context: str = "") -> None:
        """Initialize the error.

        Args:
            tag_name: The unhandled tag name
            context: Additional context about where the element was found
        """
        self.tag_name = tag_name
        msg = f"No handler for element <{tag_name}>"
        if context:
            msg = f"{msg} in {context}"
        super().__init__(msg)


class ParseEngine:
    """Engine that orchestrates element parsing using the registry.

    The engine walks the XML tree and dispatches elements to their
    registered handlers. It raises UnknownElementError for any element
    that has no handler and is not marked as skip.
    """

    def __init__(self, registry: ElementRegistry) -> None:
        """Initialize the engine with a registry.

        Args:
            registry: The element registry to use for handler lookup
        """
        self._registry = registry

    def parse(
        self,
        elem: etree._Element | None,
        context: ParseContext,
    ) -> ParseResult:
        """Parse an element tree recursively.

        Args:
            elem: The XML element to parse (or None)
            context: Current parsing context

        Returns:
            ParseResult containing the extracted text

        Raises:
            UnknownElementError: If an element has no handler and is not skipped
        """
        if elem is None:
            return ParseResult(text="")

        tag_name = get_tag_name(elem)

        # Skip marked elements
        if self._registry.should_skip(tag_name):
            return ParseResult(text="")

        # Get handler
        handler = self._registry.get_handler(elem, context)

        if handler is not None:
            # Handler exists - let it process
            def recurse(
                child: etree._Element,
                ctx: ParseContext,
            ) -> ParseResult:
                return self.parse(child, ctx)

            return handler.handle(elem, context, recurse)

        # No handler - raise error
        raise UnknownElementError(tag_name)

    def parse_children(
        self,
        elem: etree._Element,
        context: ParseContext,
        separator: str = "\n\n",
    ) -> ParseResult:
        """Parse all children of an element.

        Utility method for handlers that need to process all children.

        Args:
            elem: The parent element
            context: Current parsing context
            separator: String to join child results with

        Returns:
            ParseResult with joined text from all children
        """
        parts: list[str] = []

        for child in elem:
            result = self.parse(child, context)
            if result.text:
                parts.append(result.text)

        return ParseResult(text=separator.join(parts))

File: packages/harvester/src/registry/engine.rs
//! Parse engine that orchestrates element parsing using the registry.

use roxmltree::Node;

use super::core::ElementRegistry;
use super::types::{ParseContext, ParseResult};
use crate::error::{HarvesterError, Result};
use crate::xml::get_tag_name;

/// Engine that orchestrates element parsing using the registry.
///
/// The engine walks the XML tree and dispatches elements to their
/// registered handlers. It raises `UnknownElement` error for any element
/// that has no handler and is not marked as skip.
pub struct ParseEngine {
    registry: ElementRegistry,
}

impl ParseEngine {
    /// Create a new engine with the given registry.
    #[must_use]
    pub fn new(registry: ElementRegistry) -> Self {
        Self { registry }
    }

    /// Get a reference to the underlying registry.
    #[must_use]
    pub fn registry(&self) -> &ElementRegistry {
        &self.registry
    }

    /// Parse an element tree recursively.
    ///
    /// # Arguments
    /// * `node` - The XML element to parse
    /// * `context` - Current parsing context
    ///
    /// # Returns
    /// `ParseResult` containing the extracted text
    ///
    /// # Errors
    /// Returns `UnknownElement` error if an element has no handler and is not skipped.
    pub fn parse(&self, node: Node<'_, '_>, context: &mut ParseContext<'_>) -> Result<ParseResult> {
        let tag_name = get_tag_name(node);

        // Skip marked elements
        if self.registry.should_skip(tag_name) {
            return Ok(ParseResult::empty());
        }

        // Get handler
        if let Some(handler) = self.registry.get_handler(node, context) {
            // Create recursive parsing closure that logs errors but continues parsing
            let recurse = |child: Node<'_, '_>, ctx: &mut ParseContext<'_>| -> ParseResult {
                self.parse(child, ctx).unwrap_or_else(|err| {
                    tracing::warn!(
                        error = %err,
                        tag = %get_tag_name(child),
                        "Error parsing child element, skipping"
                    );
                    ParseResult::empty()
                })
            };

            return Ok(handler.handle(node, context, &recurse));
        }

        // No handler - raise error with parent context
        let parent_context = node
            .parent_element()
            .map(|p| format!("<{}>", get_tag_name(p)));
        Err(HarvesterError::UnknownElement {
            tag_name: tag_name.to_string(),
            context: parent_context,
        })
    }

}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::registry::handler::{ElementHandler, RecurseFn};
    use crate::registry::ElementType;
    use roxmltree::Document;

    struct TestHandler {
        output: String,
    }

    impl ElementHandler for TestHandler {
        fn element_type(&self) -> ElementType {
            ElementType::Inline
        }

        fn handle<'a, 'input>(
            &self,
            _node: Node<'a, 'input>,
            _context: &mut ParseContext<'_>,
            _recurse: &RecurseFn<'a, 'input>,
        ) -> ParseResult {
            ParseResult::new(&self.output)
        }
    }

    #[test]
    fn test_engine_parse_with_handler() {
        let mut registry = ElementRegistry::new();
        registry.register(
            "test",
            TestHandler {
                output: "hello".to_string(),
            },
        );
        let engine = ParseEngine::new(registry);

        let xml = "<test/>";
        let doc = Document::parse(xml).unwrap();
        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");

        let result = engine.parse(doc.root_element(), &mut context).unwrap();
        assert_eq!(result.text, "hello");
    }

    #[test]
    fn test_engine_parse_skip() {
        let mut registry = ElementRegistry::new();
        registry.skip(["meta-data"]);
        let engine = ParseEngine::new(registry);

        let xml = "<meta-data/>";
        let doc = Document::parse(xml).unwrap();
        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");

        let result = engine.parse(doc.root_element(), &mut context).unwrap();
        assert_eq!(result.text, "");
    }

    #[test]
    fn test_engine_parse_unknown() {
        let registry = ElementRegistry::new();
        let engine = ParseEngine::new(registry);

        let xml = "<unknown/>";
        let doc = Document::parse(xml).unwrap();
        let mut context = ParseContext::new("BWBR0000000", "2025-01-01");

        let result = engine.parse(doc.root_element(), &mut context);
        assert!(result.is_err());
    }
}
